# Deep Learning with Raw Pixel
This chapter contains the implementation and explaination of the neural network for the digit hand writing recognition. First, we begin with import the libraries as well as training and testing data. 

In [1]:
import tensorflow as tf

In [2]:
import pandas as pd
from autograd import numpy as np

In [3]:
df_train = pd.read_csv("./traindata.csv", dtype=np.uint8)

In [4]:
df_train.head(3)

id  0  1  2  3  4  5  6  7  8  ...    775  776  777  778  779  780  781  \
0   0  0  0  0  0  0  0  0  0  0  ...      0    0    0    0    0    0    0   
1   1  0  0  0  0  0  0  0  0  0  ...      0    0    0    0    0    0    0   
2   2  0  0  0  0  0  0  0  0  0  ...      0    0    0    0    0    0    0   

   782  783  label  
0    0    0      5  
1    0    0      0  
2    0    0      4  

[3 rows x 786 columns]

In [5]:
df_train.shape

(60000, 786)

In [6]:
df_imgs = df_train.drop(['label', 'id'], axis=1)
x_original = df_imgs.as_matrix().T

print("x_original shape: ", x_original.shape)

x_original shape:  (784, 60000)


/home/panitanw/anaconda3/lib/python3.6/site-packages/ipykernel_launcher.py:2: FutureWarning: Method .as_matrix will be removed in a future version. Use .values instead.
  


In [7]:
y_original = df_train['label'].as_matrix()

print("y_original shape: ", y_original.shape)

y_original shape:  (60000,)


/home/panitanw/anaconda3/lib/python3.6/site-packages/ipykernel_launcher.py:1: FutureWarning: Method .as_matrix will be removed in a future version. Use .values instead.
  """Entry point for launching an IPython kernel.


In [8]:
num_sample = 5000
inds = np.random.permutation(y_original.shape[0])[:num_sample]
x_sample = x_original[:,inds].T
y_sample = y_original[inds]

In [9]:
print("x_sample shape: ", x_sample.shape)
print("y_sample shape: ", y_sample.shape)

x_sample shape:  (5000, 784)
y_sample shape:  (5000,)


In [10]:
x = x_original.T
y = y_original

print("x shape: ", x.shape)
print("y shape: ", y.shape)

x shape:  (60000, 784)
y shape:  (60000,)


We scale the input of the algorithm by dividing the input with its maximum value. 

In [11]:
xmax = np.max(x)
x_scale = x/xmax

In [12]:
input_dim = x_scale.shape[1]
nb_classes = y.shape[0]

print("input_dim: ", input_dim)
print("nb_classes: ", nb_classes)

input_dim:  784
nb_classes:  60000


We start with a simple network containing the 3 layers: 784 input nodes, a 512-node hidden layer, and the 10-node output. We also add the 0.2 dropout to avoid overfitting. 

In [13]:
model = tf.keras.models.Sequential()
# model.add(tf.keras.layers.Flatten())
model.add(tf.keras.layers.Dense(512, 
                                input_dim=input_dim, 
                                activation=tf.nn.relu))
model.add(tf.keras.layers.Dropout(0.2))
model.add(tf.keras.layers.Dense(10, activation=tf.nn.softmax))

In [14]:
model.compile(loss='sparse_categorical_crossentropy',
              optimizer='adam', 
              metrics=['accuracy'])

Instructions for updating:
keep_dims is deprecated, use keepdims instead
Instructions for updating:
keep_dims is deprecated, use keepdims instead


Train the network with the input and its label for 5 times. 

In [15]:
model.fit(x_scale, y, epochs=5, verbose=2)

Epoch 1/5
 - 22s - loss: 0.2239 - acc: 0.9334
Epoch 2/5
 - 19s - loss: 0.0983 - acc: 0.9706
Epoch 3/5
 - 19s - loss: 0.0695 - acc: 0.9780
Epoch 4/5
 - 20s - loss: 0.0535 - acc: 0.9831
Epoch 5/5
 - 20s - loss: 0.0431 - acc: 0.9861


Read the test data, and process it. Next, we will evaluate our model with the test data.

In [16]:
df_testdata = pd.read_csv("./testdata.csv", dtype=np.uint8)
df_testdata.head(3)

id  0  1  2  3  4  5  6  7  8  ...    775  776  777  778  779  780  781  \
0   0  0  0  0  0  0  0  0  0  0  ...      0    0    0    0    0    0    0   
1   1  0  0  0  0  0  0  0  0  0  ...      0    0    0    0    0    0    0   
2   2  0  0  0  0  0  0  0  0  0  ...      0    0    0    0    0    0    0   

   782  783  label  
0    0    0      7  
1    0    0      2  
2    0    0      1  

[3 rows x 786 columns]

In [17]:
df_testimgs = df_testdata.drop(['label', 'id'], axis=1)
xtest = df_testimgs.as_matrix()

print("xtest shape:", xtest.shape)

xtest shape: (10000, 784)


/home/panitanw/anaconda3/lib/python3.6/site-packages/ipykernel_launcher.py:2: FutureWarning: Method .as_matrix will be removed in a future version. Use .values instead.
  


In [18]:
ytest = df_testdata['label'].as_matrix()
# ytest = ytest.reshape((10000, 1)).T
print("ytest shape:", ytest.shape)

ytest shape: (10000,)


/home/panitanw/anaconda3/lib/python3.6/site-packages/ipykernel_launcher.py:1: FutureWarning: Method .as_matrix will be removed in a future version. Use .values instead.
  """Entry point for launching an IPython kernel.


In [ ]:
xtest_scale = xtest/xmax

In [ ]:
model.evaluate(xtest_scale, ytest)

10000/10000 [==============================]10000/10000 [==============================] - 1s 99us/step



[0.071791607679915617, 0.97750000000000004]

As you can see from the output of the evalute method, the accuracy of the simple model is 98 percent. In this case, it is good enough to adopt this model as a classifier. However, we would like to extend this model if we achieve the perfect accuracy. The extension of this model is to add more hidden layer to the network until it almost overfit and then we start to prune it or use different technique to mitigate the overfitting problem.

In [ ]:
model = tf.keras.models.Sequential()
# model.add(tf.keras.layers.Flatten())
model.add(tf.keras.layers.Dense(input_dim, 
                                input_dim=input_dim, 
                                activation=tf.nn.relu))
model.add(tf.keras.layers.Dropout(0.2))
model.add(tf.keras.layers.Dense(nb_classes, activation=tf.nn.softmax))

model.compile(loss='sparse_categorical_crossentropy',
              optimizer='adam', 
              metrics=['accuracy'])

In [ ]:
model.fit(x_scale, y, epochs=10, verbose=2)

Epoch 1/10


In [ ]:
model.evaluate(xtest_scale, ytest)

aaa

In [ ]:
model = tf.keras.models.Sequential()
model.add(tf.keras.layers.Flatten())
model.add(tf.keras.layers.Dense(input_dim, 
                                input_dim=input_dim, 
                                activation=tf.nn.relu))
model.add(tf.keras.layers.Dropout(0.2))
model.add(tf.keras.layers.Dense(input_dim,
                                activation=tf.nn.relu))
model.add(tf.keras.layers.Dropout(0.2))
model.add(tf.keras.layers.Dense(input_dim,
                                activation=tf.nn.relu))
model.add(tf.keras.layers.Dropout(0.2))
model.add(tf.keras.layers.Dense(nb_classes, activation=tf.nn.softmax))

model.compile(loss='sparse_categorical_crossentropy',
              optimizer='adam', 
              metrics=['accuracy'])

In [ ]:
model.fit(x_scale, y, epochs=20, verbose=2)

In [ ]:
model.evaluate(xtest_scale, ytest)